src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,260 @@
+
+public class MultiplexingMiddleware : OcelotMiddleware
 {
-    public class MultiplexingMiddleware : OcelotMiddleware
+    private readonly RequestDelegate _next;
+    private readonly IResponseAggregatorFactory _factory;
+
+    public MultiplexingMiddleware(RequestDelegate next,
+        IOcelotLoggerFactory loggerFactory,
+        IResponseAggregatorFactory factory
+    )
+        : base(loggerFactory.CreateLogger<MultiplexingMiddleware>())
     {
-        private readonly RequestDelegate _next;
-        private readonly IResponseAggregatorFactory _factory;
-
-        public MultiplexingMiddleware(
-            RequestDelegate next,
-            IOcelotLoggerFactory loggerFactory,
-            IResponseAggregatorFactory factory)
-            : base(loggerFactory.CreateLogger<MultiplexingMiddleware>())
+        _factory = factory;
+        _next = next;
+    }
+
+    public async Task Invoke(HttpContext httpContext)
+    {
+        var downstreamRouteHolder = httpContext.Items.DownstreamRouteHolder();
+        var route = downstreamRouteHolder.Route;
+        var downstreamRoutes = route.DownstreamRoute;
+
+        // case 1: if websocket request or single downstream route
+        if (ShouldProcessSingleRoute(httpContext, downstreamRoutes))
         {
-            _factory = factory;
-            _next = next;
+            await ProcessSingleRoute(httpContext, downstreamRoutes.First());
+            return;
         }
 
-        public async Task Invoke(HttpContext httpContext)
+        // case 2: if no downstream routes
+        if (downstreamRoutes.Count == 0)
         {
-            var route = httpContext.Items.DownstreamRouteHolder().Route;
-            if (httpContext.WebSockets.IsWebSocketRequest)
-            {
-                // TODO: This is obviously stupid
-                httpContext.Items.UpsertDownstreamRoute(route.DownstreamRoute[0]);
-                await _next.Invoke(httpContext);
-                return;
-            }
+            return;
+        }
 
-            // Don't do anything extra if downstream route is single
-            if (route.DownstreamRoute.Count == 1)
-            {
-                httpContext.Items.UpsertDownstreamRoute(route.DownstreamRoute[0]);
-                var singleResponse = await Fire(httpContext, _next);
-                MapNotAggregate(httpContext, singleResponse);
-                return;
-            }
+        // case 3: if multiple downstream routes
+        var routeKeysConfigs = route.DownstreamRouteConfig;
+        if (routeKeysConfigs == null || !routeKeysConfigs.Any())
+        {
+            await ProcessRoutes(httpContext, route);
+            return;
+        }
 
-            if (route.DownstreamRouteConfig?.Any() != true)
-            {
-                var tasks = new Task<HttpContext>[route.DownstreamRoute.Count];
+        // case 4: if multiple downstream routes with route keys
+        var mainResponse = await ProcessMainRoute(httpContext, downstreamRoutes.First());
```

Comment:
```suggestion
        var mainResponse = await ProcessMainRoute(httpContext, downstreamRoutes[0]);
```

Reply:
Ray loves arrays! 🤩

Reply:
Accessed by the index (vs linq first) is much more efficient (even if I think the compiler optimizes it by itself)